Cannot compile with vello backend

`./mach build --feature vello`
```rust
error[E0599]: no method named `deref_mut` found for struct `GenericSharedMemory` in the current scope
   --> components\canvas\vello_backend.rs:595:79
    |
595 |                     pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
    |                                                                               ^^^^^^^^^
    |
help: there is a method `deref` with a similar name
    |
595 -                     pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
595 +                     pixels::generic_transform_inplace::<1, false, false>(data.deref());
    |

error[E0599]: no method named `to_vec` found for struct `pixels::Snapshot` in the current scope
   --> components\canvas\vello_backend.rs:631:25
    |
631 |         self.snapshot().to_vec(None, None).0
    |                         ^^^^^^ method not found in `pixels::Snapshot`

error[E0599]: no method named `to_vec` found for struct `pixels::Snapshot` in the current scope
   --> components\canvas\vello_backend.rs:635:33
    |
635 |         let (data, _, _) = data.to_vec(
    |                            -----^^^^^^ method not found in `pixels::Snapshot`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `servo-canvas` (lib) due to 3 previous errors
```

--- a/components/canvas/vello_backend.rs
+++ b/components/canvas/vello_backend.rs
@@ -589,12 +589,9 @@ impl GenericDrawTarget for VelloDrawTarget {
                 flags: ImageDescriptorFlags::empty(),
             };
             let data = SerializableImageData::Raw(if let Some(data) = data {
-                let mut data = GenericSharedMemory::from_bytes(data);
-                #[expect(unsafe_code, reason = "comply with Safety contract of deref_mut")]
-                unsafe {
-                    pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
-                }
-                data
+                GenericSharedMemory::from_bytes_with_mutator(data, |bytes| {
+                    pixels::generic_transform_inplace::<1, false, false>(bytes);
+                })
             } else {
                 GenericSharedMemory::from_byte(0, size.area() as usize * 4)
             });

--- a/components/canvas/vello_backend.rs
+++ b/components/canvas/vello_backend.rs
@@ -14,6 +14,7 @@
 use std::cell::RefCell;
 use std::collections::HashMap;
 use std::num::NonZeroUsize;
+use std::ops::DerefMut;
 use std::rc::Rc;
 
 use euclid::default::{Point2D, Rect, Size2D, Transform2D};
@@ -589,9 +590,9 @@ impl GenericDrawTarget for VelloDrawTarget {
                 flags: ImageDescriptorFlags::empty(),
             };
             let data = SerializableImageData::Raw(if let Some(data) = data {
-                let mut data = data.to_vec();
-                pixels::generic_transform_inplace::<1, false, false>(&mut data);
-                GenericSharedMemory::from_bytes(&data)
+                let mut data = GenericSharedMemory::from_bytes(data);
+                pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
+                data
             } else {
                 GenericSharedMemory::from_byte(0, size.area() as usize * 4)
             });
@@ -628,7 +629,13 @@ impl GenericDrawTarget for VelloDrawTarget {
         self.snapshot().as_raw_bytes().to_vec()
     }
 
-    fn create_source_surface_from_data(&self, data: Snapshot) -> Option<Vec<u8>> {
+    fn create_source_surface_from_data(&self, mut data: Snapshot) -> Option<Vec<u8>> {
+        data.transform(
+            SnapshotAlphaMode::Transparent {
+                premultiplied: false,
+            },
+            SnapshotPixelFormat::RGBA,
+        );
         Some(data.as_raw_bytes().to_vec())
     }
 }

--- a/components/canvas/vello_backend.rs
+++ b/components/canvas/vello_backend.rs
@@ -589,12 +589,9 @@ impl GenericDrawTarget for VelloDrawTarget {
                 flags: ImageDescriptorFlags::empty(),
             };
             let data = SerializableImageData::Raw(if let Some(data) = data {
-                let mut data = GenericSharedMemory::from_bytes(data);
-                #[expect(unsafe_code)]
-                unsafe {
-                    pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
-                };
-                data
+                let mut data = data.to_vec();
+                pixels::generic_transform_inplace::<1, false, false>(&mut data);
+                GenericSharedMemory::from_bytes(&data)
             } else {
                 GenericSharedMemory::from_byte(0, size.area() as usize * 4)
             });
@@ -628,17 +625,11 @@ impl GenericDrawTarget for VelloDrawTarget {
     }
 
     fn surface(&mut self) -> Vec<u8> {
-        self.snapshot().to_vec(None, None).0
+        self.snapshot().as_raw_bytes().to_vec()
     }
 
     fn create_source_surface_from_data(&self, data: Snapshot) -> Option<Vec<u8>> {
-        let (data, _, _) = data.to_vec(
-            Some(SnapshotAlphaMode::Transparent {
-                premultiplied: false,
-            }),
-            Some(SnapshotPixelFormat::RGBA),
-        );
-        Some(data)
+        Some(data.as_raw_bytes().to_vec())
     }
 }
 

--- a/components/canvas/vello_backend.rs
+++ b/components/canvas/vello_backend.rs
@@ -14,7 +14,6 @@
 use std::cell::RefCell;
 use std::collections::HashMap;
 use std::num::NonZeroUsize;
-use std::ops::DerefMut;
 use std::rc::Rc;
 
 use euclid::default::{Point2D, Rect, Size2D, Transform2D};
@@ -591,7 +590,10 @@ impl GenericDrawTarget for VelloDrawTarget {
             };
             let data = SerializableImageData::Raw(if let Some(data) = data {
                 let mut data = GenericSharedMemory::from_bytes(data);
-                pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
+                #[expect(unsafe_code, reason = "comply with Safety contract of deref_mut")]
+                unsafe {
+                    pixels::generic_transform_inplace::<1, false, false>(data.deref_mut());
+                }
                 data
             } else {
                 GenericSharedMemory::from_byte(0, size.area() as usize * 4)

--- a/components/shared/base/generic_channel/shared_memory.rs
+++ b/components/shared/base/generic_channel/shared_memory.rs
@@ -79,18 +79,22 @@ impl GenericSharedMemory {
         }
     }
 
-    #[expect(unsafe_code)]
-    /// # Safety
-    /// The caller must ensure that no other process or thread is reading
-    /// or writing the same data simultaneously.
-    pub unsafe fn deref_mut(&mut self) -> &mut [u8] {
-        match &mut self.0 {
+    pub fn from_bytes_with_mutator(bytes: &[u8], mutator: impl FnOnce(&mut [u8])) -> Self {
+        let mut shared_memory = Self::from_bytes(bytes);
+        match &mut shared_memory.0 {
             GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => {
-                unsafe { ipc_shared_memory.deref_mut() }
+                #[expect(unsafe_code)]
+                unsafe {
+                    mutator(ipc_shared_memory.deref_mut())
+                }
             },
-            GenericSharedMemoryVariant::InProcess(arc) =>
-                Arc::get_mut(arc).expect("User should uphold safety contract: no other reference to the same data should exist").as_mut_slice(),
+            GenericSharedMemoryVariant::InProcess(arc) => mutator(
+                Arc::get_mut(arc)
+                    .expect("Arc just created from bytes")
+                    .as_mut_slice(),
+            ),
         }
+        shared_memory
     }
 }
 

--- a/components/shared/base/generic_channel/shared_memory.rs
+++ b/components/shared/base/generic_channel/shared_memory.rs
@@ -3,7 +3,7 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
 use std::fmt;
-use std::ops::{Deref, DerefMut};
+use std::ops::Deref;
 use std::sync::Arc;
 
 use ipc_channel::ipc::IpcSharedMemory;
@@ -29,19 +29,6 @@ impl Deref for GenericSharedMemory {
         match &self.0 {
             GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => ipc_shared_memory,
             GenericSharedMemoryVariant::InProcess(items) => items.as_slice(),
-        }
-    }
-}
-
-impl DerefMut for GenericSharedMemory {
-    fn deref_mut(&mut self) -> &mut [u8] {
-        match &mut self.0 {
-            GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => {
-                #[expect(unsafe_code)]
-                unsafe { ipc_shared_memory.deref_mut() }
-            },
-            GenericSharedMemoryVariant::InProcess(arc) =>
-                Arc::get_mut(arc).expect("User should uphold safety contract: no other reference to the same data should exist").as_mut_slice(),
         }
     }
 }
@@ -89,6 +76,20 @@ impl GenericSharedMemory {
                 Arc::new(ipc_shared_memory.to_vec())
             },
             GenericSharedMemoryVariant::InProcess(arc) => arc,
+        }
+    }
+
+    #[expect(unsafe_code)]
+    /// # Safety
+    /// The caller must ensure that no other process or thread is reading
+    /// or writing the same data simultaneously.
+    pub unsafe fn deref_mut(&mut self) -> &mut [u8] {
+        match &mut self.0 {
+            GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => {
+                unsafe { ipc_shared_memory.deref_mut() }
+            },
+            GenericSharedMemoryVariant::InProcess(arc) =>
+                Arc::get_mut(arc).expect("User should uphold safety contract: no other reference to the same data should exist").as_mut_slice(),
         }
     }
 }

--- a/components/shared/base/generic_channel/shared_memory.rs
+++ b/components/shared/base/generic_channel/shared_memory.rs
@@ -3,7 +3,7 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
 use std::fmt;
-use std::ops::Deref;
+use std::ops::{Deref, DerefMut};
 use std::sync::Arc;
 
 use ipc_channel::ipc::IpcSharedMemory;
@@ -29,6 +29,19 @@ impl Deref for GenericSharedMemory {
         match &self.0 {
             GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => ipc_shared_memory,
             GenericSharedMemoryVariant::InProcess(items) => items.as_slice(),
+        }
+    }
+}
+
+impl DerefMut for GenericSharedMemory {
+    fn deref_mut(&mut self) -> &mut [u8] {
+        match &mut self.0 {
+            GenericSharedMemoryVariant::Ipc(ipc_shared_memory) => {
+                #[expect(unsafe_code)]
+                unsafe { ipc_shared_memory.deref_mut() }
+            },
+            GenericSharedMemoryVariant::InProcess(arc) =>
+                Arc::get_mut(arc).expect("User should uphold safety contract: no other reference to the same data should exist").as_mut_slice(),
         }
     }
 }

